kernel/src/client/arrow_utils.rs

```diff
@@ -0,0 +1,78 @@
+//! Some utilities for working with arrow data types
+
+use std::sync::Arc;
+
+use crate::{DeltaResult, Error};
+
+use arrow_array::RecordBatch;
+use arrow_schema::{Schema as ArrowSchema, SchemaRef as ArrowSchemaRef};
+use itertools::Itertools;
+use parquet::{arrow::ProjectionMask, schema::types::SchemaDescriptor};
+
+/// Get the indicies in `parquet_schema` of the specified columns in `requested_schema`
+pub(crate) fn get_requested_indicies(
+    requested_schema: &ArrowSchema,
+    parquet_schema: &ArrowSchemaRef,
+) -> DeltaResult<Vec<usize>> {
+    let indicies = requested_schema
+        .fields
+        .iter()
+        .map(|field| {
+            // todo: handle nested (then use `leaves` not `roots` below in generate_mask)
+            parquet_schema.index_of(field.name())
+        })
+        .try_collect()?;
+    Ok(indicies)
+}
+
+/// Create a mask that will only select the specified indicies from the parquet. Currently we only
+/// handle "root" level columns, and hence use `ProjectionMask::roots`, but will support leaf
+/// selection in the future
+pub(crate) fn generate_mask(
+    requested_schema: &ArrowSchema,
+    parquet_schema: &ArrowSchemaRef,
+    parquet_physical_schema: &SchemaDescriptor,
+    indicies: &[usize],
+) -> Option<ProjectionMask> {
+    if parquet_schema.fields.size() == requested_schema.fields.size() {
+        // we assume that in get_requested_indicies we will have caught any column name mismatches,
+        // so here we can just say that if we request the same # of columns as the parquet file
+        // actually has, we don't need to mask anything out
+        None
+    } else {
+        Some(ProjectionMask::roots(
+            parquet_physical_schema,
+            indicies.to_owned(),
+        ))
+    }
+}
+
+/// Reorder a RecordBatch to match `requested_schema`. This method takes `indicies` as computed by
+/// [`get_requested_indicies`] as an optimization. If the indicies are in order, then we don't need
+/// to do any re-ordering.
```

Comment:
We read parquet in two ways:
1. Using [ParquetRecordBatchStreamBuilder](https://arrow.apache.org/rust/parquet/arrow/async_reader/type.ParquetRecordBatchStreamBuilder.html) in the default client
2. Using [ParquetRecordBatchReaderBuilder](https://arrow.apache.org/rust/parquet/arrow/arrow_reader/type.ParquetRecordBatchReaderBuilder.html) in the sync client

They are basically the same but sync/async.

Neither have a `with_schema` type option. The only way to potentially pass a schema that I could see was via passing an [ArrowReaderMetadata](https://arrow.apache.org/rust/parquet/arrow/arrow_reader/struct.ArrowReaderMetadata.html) to a `new_with_metadata` constructor. I dug into the code a bit and it's not clear if that will actually do anything, but I can test it out.

The reason I assumed it _doesn't_ work is that it actually does not handle a "pruned" schema. You instead just give it a `ProjectionMask` which it applies over the full schema (see eg [here](https://arrow.apache.org/rust/parquet/arrow/arrow_reader/type.ParquetRecordBatchReaderBuilder.html#method.with_projection)). It uses similar techniques for filtering rows and row groups. I assumed they would have had a way to specify the ordering of the columns you wanted as well with some kind of similar method rather than expecting you to construct a metadata and then modify its internal schema.

Reply:
Actually, there's just no way I can see to do it. I now recall as I went to test:
1. the `schema` field on an ArrowReaderMetadata](https://arrow.apache.org/rust/parquet/arrow/arrow_reader/struct.ArrowReaderMetadata.html) is only `pub(crate)` not `pub`, so you can't modify it
2. The only constructor that doesn't just load it from a file is [try_new](https://arrow.apache.org/rust/parquet/arrow/arrow_reader/struct.ArrowReaderMetadata.html#method.try_new) 
3. That takes a `FileMetaData` which does indeed have a schema inside of it, but from digging around the API it's pretty clear you're not supposed to modify it. I think you could perhaps pull all the fields out and reorder/filter them, but I suspect that would just cause an error, and the code would be hairy.